lucet-runtime/lucet-runtime-internals/src/region/uffd.rs

```diff
@@ -0,0 +1,512 @@
+                                )
+                                .map_err(|e| Error::InternalError(e.into()))?;
+                            }
+                        } else {
+                            // else if outside the sparse data area, or with an empty page
+                            // eprintln!("zeroing a page at {:p}", fault_page as *mut c_void);
+                            unsafe {
+                                uffd.zeropage(fault_page as *mut c_void, host_page_size(), true)
+                                    .map_err(|e| Error::InternalError(e.into()))?;
+                            }
+                        }
+                    }
+                }
+            }
+            Ok(Some(ev)) => panic!("unexpected uffd event: {:?}", ev),
+        }
+    }
+
+    Ok(())
+}
+
+impl Region for UffdRegion {
+    fn free_slots(&self) -> usize {
+        self.freelist.lock().unwrap().len()
+    }
+
+    fn used_slots(&self) -> usize {
+        self.capacity() - self.free_slots()
+    }
+
+    fn capacity(&self) -> usize {
+        self.instance_capacity
+    }
+}
+
+impl RegionInternal for UffdRegion {
+    fn new_instance_with(
+        &self,
+        module: Arc<dyn Module>,
+        embed_ctx: CtxMap,
+        heap_memory_size_limit: usize,
+    ) -> Result<InstanceHandle, Error> {
+        let custom_limits;
+        let mut limits = self.get_limits();
+
+        // Affirm that the module, if instantiated, would not violate
+        // any runtime memory limits.
+        match heap_memory_size_limit.cmp(&limits.heap_memory_size) {
+            Ordering::Less => {
+                // The supplied heap_memory_size is smaller than
+                // default. Augment the limits with this custom value
+                // so that it may be validated.
+                custom_limits = Limits {
+                    heap_memory_size: heap_memory_size_limit,
+                    ..*limits
+                };
+                limits = &custom_limits;
+            }
+            Ordering::Equal => (),
+            Ordering::Greater => {
+                return Err(Error::InvalidArgument(
+                    "heap memory size requested for instance is larger than slot allows",
+                ))
+            }
+        }
+        module.validate_runtime_spec(&limits)?;
+
+        let slot = self
+            .freelist
+            .lock()
+            .unwrap()
+            .pop()
+            .ok_or(Error::RegionFull(self.instance_capacity))?;
+
+        if slot.heap as usize % host_page_size() != 0 {
+            lucet_bail!("heap is not page-aligned; this is a bug");
+        }
+
+        let limits = &slot.limits;
+        module.validate_runtime_spec(limits)?;
+
+        for (ptr, len) in [
+            // zero the globals
+            (slot.globals, limits.globals_size),
+            // zero the sigstack
+            (slot.sigstack, limits.signal_stack_size),
+        ]
+        .into_iter()
```

Comment:
I am new to userfaultfd, and I'm wondering about a small difference between its implementation of new_instance_with and the mmap implementation.  https://github.com/bytecodealliance/lucet/blob/83fd2248b36c780cc74c0abb49024cca4a0c053c/lucet-runtime/lucet-runtime-internals/src/region/mmap.rs#L121-L122

Reply:
So yeah, that's intentional here. We're making the allocation of the stack lazy. If you look at the other lines changed in that commit, you'll see where I make page faults in the Stack area zero out the page on demand. 

That said, it does point to a lack of documentation here that I should fix.

Reply:
Added some docs!